Switch pygeofilter dependency to PyPI package

JBurkinshaw opened the pull request:
# Description

Switch pygeofilter dependency to PyPI package

- Remove `pygeofilter` from `requirements_dev.txt`.
- Delete `requirements_build.txt`, remove pip install from dockerfile and add dependencies to `setup.py` instead.
- Remove getter build stage which downloaded pygeofilter source.
- Remove pygeofilter pip install to stop install from source in build.
- Add version to pygeofilter in setup.py.
- Update README.md

Fixes #4 

## Type of change

Please delete options that are not relevant.

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] This change requires a documentation update

# Instructions

Running the test suite without errors will confirm this change is effective.

# How Has This Been Tested?

- [x] Run `scripts/test`


# Checklist:

- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have made corresponding changes to the documentation
- [x] My changes generate no new warnings
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] New and existing unit tests pass locally with my changes
- [x] Any dependent changes have been merged and published in downstream modules

JBurkinshaw committed:
Switch pygeofilter dependency to PyPI package
Remove pygeofilter from requirements_dev.txt
pygeofilter trequires Alpine build step - add to requirements_build.txt.
Remove getter build stage which downloaded pygeofilter source.
Remove pygeofilter pip install to stop install from source in build.
Add version to pygeofilter in setup.py.

JBurkinshaw committed:
Update README

JBurkinshaw commented on feapi/app/setup.py:
```diff
@@ -7,7 +7,7 @@
 inst_reqs = [
     "pydantic==1.8.2",
-    "pygeofilter",
+    "pygeofilter==0.0.2",
```
It is not clear to me if `pygeofilter` is required here. An OS dependency is required to build with an Alpine docker base image so it is also included in `requirements_build.txt`.
`scripts/test` is successful both with and without pygeofilter here.

JBurkinshaw committed:
Alternate python:3.8-slim base image

JBurkinshaw committed:
Update testing Dockerfile to use apt-get
Add requests to requirements-test.txt
Remove + from 3.8+ from "Developing" section in readme

JBurkinshaw committed:
Delete requirements_build.txt
Remove pyproj from setup.py

JBurkinshaw commented on feapi/app/setup.py:
```diff
@@ -6,8 +6,13 @@
 
 inst_reqs = [
-    "pydantic==1.8.2",
-    "pygeofilter",
+    "uvicorn==0.13.4",
+    "gunicorn==20.1.0",
+    "uvloop==0.15.2",
+    "httptools==0.2.0",
+    "pygeofilter==0.0.2",
+    "psycopg2==2.8.6",
+    "asyncpg==0.23.0",
```
Added pygeofilter here along with libraries that were previously in the now deleted `requirements_build.txt`

JBurkinshaw commented on requirements.txt:
```diff
@@ -5,6 +5,7 @@ Jinja2==3.0.1
 iso8601==0.1.14
 pytz==2021.1
 aiofiles==0.7.0
+pyproj==2.6.1
```
`pyproj` is required here for the tests to pass.
Do `pyproj` or any of the other dependencies here need to be included in `setup.py` for creating a python package in the future?

cuttlefish replied:
There seems to be one in `feapi/app/`
I think it'll take some experimentation to figure out what the minimum requirements are.

cuttlefish commented on Dockerfile:
```diff
@@ -1,28 +1,11 @@
-RUN git checkout 5cfa8afbb6ffc8a30b8dde5be7a2981d4b534678
+FROM python:3.8-slim
 
+RUN apt-get update && apt-get -y install python3-pip python3-psycopg2 libpq-dev
+RUN rm -rf /var/lib/apt/lists/*
```
Docker uses a diff based file system. A "layer" gets created after each command in a Dockerfile.
This `rm` will remove the files listed, but the files were written to the previous layer. So the final image size will be the same (technically adding the `rm` makes it a very small amount larger).
To get the desired effect here, the two `RUN` statements need to be combined.

```suggestion
RUN apt-get update && apt-get install -y \
        libpq-dev \
        python3-pip \
        python3-psycopg2 \
    && rm -rf /var/lib/apt/lists/*
```

cuttlefish commented on Dockerfile:
```diff
@@ -1,28 +1,11 @@
-RUN apk add --no-cache --update \
-  bash \
-  curl
+RUN apt-get update && apt-get install -y curl
```
```suggestion
RUN apt-get update && apt-get install -y curl \
    && rm -rf /var/lib/apt/lists/*
```

JBurkinshaw replied:
Unless I'm misunderstanding something, I think the two `apt-get` install commands can be combined:
```dockerfile
RUN apt-get update && apt-get install -y \
        libpq-dev \
        python3-pip \
        python3-psycopg2 \
        curl \
    && rm -rf /var/lib/apt/lists/*
```

cuttlefish commented on feapi/testing/Dockerfile:
```diff
@@ -1,6 +1,7 @@
 FROM feapi
 
-RUN apk add --no-cache --update postgresql-client
+RUN apt-get update && apt-get install -y postgresql-client-11
+RUN rm -rf /var/lib/apt/lists/*
```
```suggestion
RUN apt-get update && apt-get install -y postgresql-client-11 \
    && rm -rf /var/lib/apt/lists/*
```

cuttlefish commented on feapi/testing/Dockerfile:
```diff
@@ -1,6 +1,7 @@
+RUN apt-get update && apt-get install -y postgresql-client-11
+RUN rm -rf /var/lib/apt/lists/*
 
 COPY requirements-test.txt /requirements-test.txt
 RUN pip install -r /requirements-test.txt
```
I usually try not to put file in `/`. It can lead to confusion down the road.
I'd set a `WORKDIR` and do the `COPY` and `RUN` relative to that.
I like `/opt/$NAME` but there are lots of options there.

cuttlefish reviewed (changes requested):
Made some suggestions on Dockerfile structure

cuttlefish commented:
@JBurkinshaw The testing Dockerfile seems to inherit from the main image, then install a bunch of dependencies. That'll be pretty slow. All of the test requirements will need to be installed every time a single file changes in the base image.
A multistage build would be a lot faster, but I'm not sure how much time you have to put into this.

cuttlefish commented on Dockerfile:
```diff
@@ -1,28 +1,11 @@
+RUN apt-get update && apt-get install -y curl
 
 COPY requirements.txt /requirements.txt
 RUN pip install -r /requirements.txt
```
`pip install --no-cache-dir` will save you some image size

JBurkinshaw replied:
> @JBurkinshaw The testing Dockerfile seems to inherit from the main image, then install a bunch of dependencies. That'll be pretty slow. All of the test requirements will need to be installed every time a single file changes in the base image.
> A multistage build would be a lot faster, but I'm not sure how much time you have to put into this.

Putting this in another issue to address later: #19

JBurkinshaw commented:
> @JBurkinshaw The testing Dockerfile seems to inherit from the main image, then install a bunch of dependencies. That'll be pretty slow. All of the test requirements will need to be installed every time a single file changes in the base image.
> A multistage build would be a lot faster, but I'm not sure how much time you have to put into this.

Putting this in another issue to address later: #19

JBurkinshaw committed:
Changes based on PR feedback
Combine apt-get commands into single RUN statement.
Copy requirements files to /opt/$NAME instead of root.

JBurkinshaw commented on Dockerfile:
```diff
@@ -1,33 +1,21 @@
-RUN git checkout 5cfa8afbb6ffc8a30b8dde5be7a2981d4b534678
+FROM python:3.8-slim
 
+RUN apt-get update && apt-get install -y \
+      # OS dependencies only required to build certain Python dependencies
+      libpq-dev \
+      python3-pip \
+      python3-psycopg2 \
+      # OS dependency required at runtime
+      curl \ 
+  && rm -rf /var/lib/apt/lists/*
```
Decided to combine all of the dependencies (build and runtime) here for now. Created an issue to enhance the dockerfiles to benefit from multistage build in the future: #19

JBurkinshaw committed:
Merge branch 'main' into feature/pygeofilter-pypi
Incorporate switch from feapi to oaff terminology.

cuttlefish commented on Dockerfile:
```diff
@@ -1,33 +1,21 @@
-COPY requirements.txt /requirements.txt
-RUN pip install -r /requirements.txt
-RUN rm /requirements.txt
+WORKDIR /opt/$NAME
+COPY requirements.txt requirements.txt
+RUN pip install -r requirements.txt --no-cache-dir
+RUN rm requirements.txt
 
+WORKDIR /
 COPY ./oaff /oaff
 RUN pip install -e /oaff/app
```
Sorry, I was very unclear there and shouldn't have used plausible looking syntax. I meant something like this:
(`ogc-api-fast-features` is pretty unwieldy, but so is `/opt/oaff/oaff`)
```suggestion
WORKDIR /opt/ogc-api-fast-features
COPY requirements.txt ./
RUN pip install --no-cache-dir -r requirements.txt

COPY ./oaff oaff
RUN pip install -e oaff/app
```

cuttlefish commented on oaff/testing/Dockerfile:
```diff
@@ -1,10 +1,15 @@
 
-COPY requirements-test.txt /requirements-test.txt
-RUN pip install -r /requirements-test.txt
+WORKDIR /opt/$NAME
+COPY requirements-test.txt requirements-test.txt
+RUN pip install -r requirements-test.txt --no-cache-dir 
 
+WORKDIR /
 RUN mkdir -p /oaff/testing
 COPY data /oaff/testing/data
```
If you implement my suggestion above, `WORKDIR` will already be set to `/opt/ogc-api-fast-features`
```suggestion
WORKDIR /opt/$NAME
COPY requirements-test.txt ./
RUN pip install --no-cache-dir -r requirements-test.txt

RUN mkdir -p oaff/testing
COPY data oaff/testing/data
```

cuttlefish commented on oaff/testing/Dockerfile:
```diff
@@ -1,10 +1,15 @@
 COPY data /oaff/testing/data
-COPY integration_tests /oaff/testing/integration_tests
\ No newline at end of file
+COPY integration_tests /oaff/testing/integration_tests
```
```suggestion
COPY integration_tests oaff/testing/integration_tests
```

cuttlefish reviewed (changes requested):
Sorry about that, I was pretty unclear in my previous comments and included some _way_ too plausible syntax that I meant as pseudo code :facepalm:. Hopefully this is a bit more clear.
I used to help maintain some installers, so I get a bit neurotic about placement. I usually use `/opt/projectname` for stuff like this. That's still not exactly right, but it's easy and relatively short. We can leave it in `/oaff` for now, but the `$NAME` part doesn't actually work. That was just a really poor choice of syntax on my part.
For more than you could ever possibly want to know:
https://refspecs.linuxfoundation.org/FHS_3.0/fhs-3.0.html#optAddonApplicationSoftwarePackages

JBurkinshaw committed:
Update workdir to /opt/ogc-api-fast-features

JBurkinshaw committed:
Modift postgres loader service command

JBurkinshaw committed:
Update docker compose api volume dir for container

cuttlefish reviewed (approved):
This looks great. I think you've cleaned it up quite a bit.
Sorry this turned into such a big job!
